src/dodal/devices/s4_slit_gaps.py

```diff
@@ -1,8 +1,58 @@
-from ophyd import Component, Device, EpicsMotor
+from ophyd_async.core import DeviceVector, StandardReadable
+from ophyd_async.epics.motion import Motor
 
 
-class S4SlitGaps(Device):
-    """Note that the S4 slits have a different PV fromat to other beamline slits"""
+class S4SlitGaps(StandardReadable):
+    """
+    Representation of S4 Slit Gaps, a group of motors controlling the xray optics slits
+    """
 
-    xgap = Component(EpicsMotor, "XGAP")
-    ygap = Component(EpicsMotor, "YGAP")
+    def __init__(self, prefix: str, name: str = "") -> None:
+        self.x_gap = Motor(prefix + "X:SIZE")
+        self.y_gap = Motor(prefix + "Y:SIZE")
+        self.x_centre = Motor(prefix + "X:CENTRE")
+        self.y_centre = Motor(prefix + "Y:CENTRE")
+
+        # Type ignore pending: https://github.com/bluesky/ophyd-async/issues/192
+        self.set_readable_signals(
+            read=[
+                self.x_gap,
+                self.y_gap,
+                self.x_centre,
+                self.y_centre,
+            ]  # type: ignore
+        )
+        super().__init__(name)
+
+
+class S4SlitGapsGroup(StandardReadable):
+    """
+    A group of slits, since they normally come in groups following a standard PV naming
+    convention.
+    """
+
+    def __init__(
+        self,
+        prefix: str,
+        indices: range,
+        name: str = "",
+    ) -> None:
+        """
+        Constructor, formats prefix with indices for sub-devices. For example:
+        S4SlitGapsGroup("MY-SLITS-{0:02d}:", range(5)) will populate the group with 4
+        S4SlitGaps devices with prefixes MY-SLITS-00, MY-SLITS-01, MY-SLITS-02,
+        MY-SLITS-03. The index to access them will be the same as the PV number, so
+        MY-SLITS-02 is accessed via group.slits[2].
```

Comment:
I think at the very least, indices should be a Sequence[int], in case Slits[n] are disabled temporarily.

Looks like this fits with the ophyd_async ADR https://github.com/bluesky/ophyd-async/blob/7be73e3ce0f79fca2e02f77f366792ee84586ed7/docs/developer/explanations/decisions/0006-procedural-device-definitions.rst#L58 for proceedurally created devices.

I do wish it was just slits[1] rather than group.slits[1]: does group.name = "" work with blueapi still being able to get group.slits[1]?

Reply:
Agree with `Sequence[int]`, for why the group adds another layer to the tree, see this comment https://github.com/bluesky/ophyd-async/issues/181#issuecomment-2045537823

That said, I do agree it would be nicer, I wonder if there's a way of subclassing `DeviceVector` to incorporate the PV prefixing logic...

Tagging @coretl in case he has ideas

Reply:
Personally I'm not a fan of a composite device here, I'd prefer `slits1`, `slits2`, etc.

However, if you decide you really want it, it does occur to me that multiple inheritance of `DeviceVector` and `StandardDetector` (rather than your existing nesting of `DeviceVector` within `StandardDetector` might work to squash down the layers

Reply:
Makes sense and I agree with the user perspective. @DominicOram how about `i03.slits()[1]`?

Reply:
I think I'd also prefer `i22.slits1()` or `i22.slits_1()`.

The slit sets are conceptually independent from each other, and it does happen that a slit set gets removed for various reasons.

Reply:
Makes sense, I think we just remove the group entirely if it doesn't make semantic sense

Reply:
Although this does create a lot of additional boilerplate in `i22.py` so should think more about it long-term

Reply:
I'm not opposed to the boilerplate in this case. To be honest, long term we're probably going to need to split `i22` into modules like `i22.slits`, `i22.sample_environment`, and so on, and the top-level `i22` will just `import *` from each of those.

Reply:
I think I fear the boilerplate just because of how much GDA's per-beamline config has ballooned over the years, it's ended up split into files and subdirectories etc. in a very similar way to your suggestion here. I suppose that's not the end of the world but it would be nice to have a more concise description of the beamline to work with.